### ISSUE TITLE
Retrieve the current firewall name for a Request

### ISSUE THREAD

**Issue by trsteel88** [symfony/symfony#14435]
I am trying to determine the current firewall name for a Request (application has multiple firewalls). Is this possible?

I can see a map is creating in [SecurityExtension.php](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L242)

However, I can't use this map as it is only available to the Firewall which doesn't expose a way to get the map. Secondly, from what I can see, the firewall name isn't stored anywhere for me to retrieve this as everything references services.

I can see that a ProviderKey() is available for some tokens (e.g UsernamePasswordToken). However, this method is not on the TokenInterface so I cannot rely on it.

Basically I want to add a generic EventListener for Authentication successes/failures. It needs to store the firewall name so it can be determined which firewall a user was trying to login to.

**Comment by hacfi** [comment]
@trsteel88 Just debugged this issue and I think you’re right..the firewall name is not accessible directly.

If you urgently need this you can create a very ugly hack:
Add a compiler pass which "steals" the security firewall map:

``` php
        $map = $container->getDefinition('security.firewall.map');

        $maps = $map->getArgument(1);

        $refs = [];

        foreach ($maps as $serviceName => $firewall) {
            $refs[substr($serviceName, 30)] = $firewall;
        }
```

then you add `$refs` to a service which contains the same logic as `Symfony\Bundle\SecurityBundle\Security\FirewallMap`: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php#L38-42

and make the result available.


I agree that this should be exposed :).

**Comment by trsteel88** [comment]
Thanks for getting back to me @hacfi. I considered that but it felt very hacky.

I thought it may have been exposed somewhere so I was trying to avoid it. Since it isn't, I will have to use this method.

Would love to see a public method made available.

**Comment by iltar** [comment]
The new logout function of twig in 2.7 has support for a specific firewall, maybe that can point you in the right direction.

**Comment by jaytaph** [comment]
That code is actually inside the logout generator: https://github.com/symfony/symfony/blob/5bb2816b0f0300c2143f25d0cd82de2831b94abb/src/Symfony/Component/Security/Http/Logout/LogoutUrlGenerator.php#L98

It will try and fetch the providerKey from the current token. It assumes that each firewall has a providerkey that matches the firewall name (this doesn't have to be the case though, sometime you have multiple firewalls all using the same provider key).

**Comment by thomas-koch-comsolit** [comment]
Another use case: SonataUserBundle checks, whether it should render an "impersonate" button in the user admin:
https://github.com/sonata-project/SonataUserBundle/blob/master/Admin/Model/UserAdmin.php#L68

The role necessary however is configurable. So it would be necessary to retrieve not only the name of the current firewall but the configuration of the current firewall.

It would be good to have the active firewall available as a service so one could retrieve informations as necessary.

**Comment by iltar** [comment]
What about a `FirewallAwareTokenInterface` that has the `getProviderKey` method? I would vote to rename it though, provider key is very confusing because we're talking about a firewall (context) identifier.

On a side note, would this be something to store in the token attributes instead? Maybe @stof knows this.

**Comment by thomas-koch-comsolit** [comment]
I think more useful (and consistent with the name `FirewallAwareTokenInterface`) would be a method `getFirewall` that returns an object representing the firewall that can be questioned. That's easier than trying to lookup all informations from the configuration by the providerKey.

**Comment by trsteel88** [comment]
fyi, this is how I am retrieving the firewall name:

Utility to retrieve firewall name:

```php
<?php

namespace Vivo\SecurityBundle\Util;

use Symfony\Component\HttpFoundation\Request;

class FirewallManager
{
    /**
     * @var \Symfony\Component\HttpFoundation\RequestMatcher[]
     */
    private $map;

    /**
     * @param \Symfony\Component\HttpFoundation\RequestMatcher[] $map
     */
    public function __construct(array $map)
    {
        $this->map = $map;
    }

    /**
     * @param  Request     $request
     * @return string|null
     */
    public function getFirewallNameForRequest(Request $request)
    {
        foreach ($this->map as $firewallName => $requestMatcher) {
            if (null === $requestMatcher || $requestMatcher->matches($request)) {
                return $firewallName;
            }
        }

        return null;
    }
}

```

Compiler pass to inject firewall map:

```php
<?php

namespace Acme\DemoBundle\DependencyInjection\Compiler;

use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;

class FirewallManagerCompilerPass implements CompilerPassInterface
{
    public function process(ContainerBuilder $container)
    {
        if (false === $container->hasAlias('acme_demo.util.firewall_manager') && false === $container->hasDefinition('acme_demo.util.firewall_manager')) {
            return;
        }

        $map = $container->getDefinition('security.firewall.map');
        $maps = $map->getArgument(1);

        $refs = [];
        foreach ($maps as $serviceName => $firewall) {
            $refs[substr($serviceName, 30)] = $firewall;
        }

        $firewallManagerDef = $container->getDefinition('acme_demo.util.firewall_manager');
        $firewallManagerDef->replaceArgument(0, $refs);
    }
}
```

**Comment by caponica** [comment]
:+1:

**Comment by javiereguiluz** [comment]
Closing it in favor of #15294, which not only would provide the name of the firewall but more information.
